### PR TITLE
[cxx-interop] Fix SIL deserialization error

### PR DESCRIPTION
**Pull request by egorzhdan**
Make sure nested namespaces are added to the SwiftLookupTable when they are declared in a non-canonical redecl of the parent namespace.

This was blocking C++ interop adoption in SwiftCompilerSources.

### CODE REVIEW

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
This is a requirement for https://github.com/apple/swift/pull/42297.

**Comment by egorzhdan** [comment]
@swift-ci please clean smoke test macOS

**Comment by egorzhdan** [comment]
Hmm that wasn't a clean build 🤔

**Comment by egorzhdan** [comment]
@swift-ci please clean smoke test macOS platform

**Comment by egorzhdan** [comment]
@swift-ci please clean test macOS platform

**Comment by egorzhdan** [comment]
@swift-ci please clean smoke test

File: test/Interop/Cxx/class/nested-records-module-interface.swift [hunk 1]
```diff
@@ -82,7 +82,7 @@
 // CHECK:   struct ForwardDeclaredFriend {
 // CHECK:     init()
 // CHECK:   }
-// CHECK:   static func takesFriend(_ b: NestedDeclIsAFirstForwardDeclaration.ForwardDeclaredFriend)
```
Comment (zoecarver) [inline test/Interop/Cxx/class/nested-records-module-interface.swift:85]: This seems odd. Any ideas why this changed?

**Review by zoecarver — APPROVED** [review]
🚢 it

File: test/Interop/Cxx/class/nested-records-module-interface.swift (hunk 1 above)
Comment (egorzhdan) [reply test/Interop/Cxx/class/nested-records-module-interface.swift:85]: Previously we went through all the decls in a namespace, and emitted them, and now we request a canonical decl for each of the decls in a namespace, to make sure we don't emit the same decl multiple times if it has redecls inside different redecls of the namespace.
Different redecls of this function have different parameter names, so I had to update this test.

**Comment by egorzhdan** [comment]
This builds successfully on Linux/Windows CI, but fails on macOS. Locally for me it builds fine on macOS. I'll need to debug this.

**Comment by egorzhdan** [comment]
@swift-ci please smoke test macOS platform

**Comment by egorzhdan** [comment]
@swift-ci please smoke test macOS platform

**Comment by egorzhdan** [comment]
@swift-ci please smoke test macOS platform

**Comment by egorzhdan** [comment]
@swift-ci please smoke test macOS platform

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
I added one more commit with a change in `ClangImporter.cpp` to fix the build on macOS, @zoecarver could you please take another look?

**Review by zoecarver — APPROVED** [review]
Send it! Thanks again, Egor.

**Pull request merged**